### ISSUE TITLE
Worker goes unhealthy after 10 minutes on sidekiq 7

### ISSUE THREAD

**Issue by w32-blaster** [arturictus/sidekiq_alive#94]
Hi,

We are trying to upgrade our infrastructure to use sidekiq 7 and facing the following issue:
sidekiq_alive works fine when registering the worker in redis, then it becomes healthy. 
After exactly 10 minutes it becomes unhealthy with this message: "Can't find the alive key" and pod gets restarted. 
We have verified and during this 10 minutes period we can see it really alive, so it's not like our health check starts working after 10 minutes. 
Any idea where to look to solve this problem?

**Comment by arturictus** [comment]
Hi @w32-blaster ,
Is the sidekiq-alive queue running? Can you check if it shows in the interface?

**Comment by w32-blaster** [comment]
@arturictus do you mean to check it in the sidekiq UI?

**Comment by arturictus** [comment]
@w32-blaster, yes

**Comment by rakibulislam** [comment]
Hi @arturictus Yes, sidekiq-alive queue is running and we can see that in the sidekiq UI. Please let us know if you have any clue on why this might be happening with this sidekiq-alive gem when used with sidekiq 7.1. Just anothre quick info: we also use `sidekiq-cron` since this issue is happening every 10 minutes, would the `sidekiq-cron` gem play any role here somehow? Please let us know if you can think of anything to solve this issue. Thanks! cc: @w32-blaster

**Comment by andrcuns** [comment]
I am also using it with sidekiq 7.1 together with `sidekiq-cron` and it seems to work fine.

One problem that can happen is if your cron jobs spawn enough jobs that take all available threads, alive job will not get picked up and app will start reporting as unhealthy. I encountered such issue a while back in my project here: https://gitlab.com/dependabot-gitlab/dependabot/-/issues/293#note_1408148243

This is not really solvable in sidekiq < 7 but capsules in the new version should be able to address that. Potential fix: https://github.com/arturictus/sidekiq_alive/pull/96

**Comment by w32-blaster** [comment]
That's not the case. Our worker threads are sitting there and doing nothing. Then after 10 minutes they got removed because the health check.

**Comment by andrcuns** [comment]
You should try and monitor the sidekiq scheduled jobs page and see if it consistently has alive job scheduled:
![image](https://github.com/arturictus/sidekiq_alive/assets/12693088/ce7ce91d-f18c-4ad1-8783-536aecefcbaf)

Given the default ttl value is 10 minutes, it basically means that sidekiq-alive is not working at all. You can change the option `config.time_to_live = 10 * 60` in the config to validate that.

**Comment by rakibulislam** [comment]
> You should try and monitor the sidekiq scheduled jobs page and see if it consistently has alive job scheduled: ![image](https://user-images.githubusercontent.com/12693088/264628272-ce7ce91d-f18c-4ad1-8783-536aecefcbaf.png)
> 
> Given the default ttl value is 10 minutes, it basically means that sidekiq-alive is not working at all. You can change the option `config.time_to_live = 10 * 60` in the config to validate that.

@andrcuns You are right. I just checked and verified this: sidekiq-alive queue is not scheduled in staging (where we deployed sidekiq 7), but it's present in other environments where we still have sidekiq 6.5.6.

Now, could you please help us a bit debug where is the actuall issue? Here are list of gems and their versions we are using along with `sidekiq 7.1.2`: `sidekiq-cron-1.10.1`, `sidekiq-failures-1.0.4`, `sidekiq-limit_fetch-4.4.1`, `sidekiq-unique-jobs-8.0.3`, `sidekiq_alive-2.2.3`. We are also using `redis-5.0.7` and `hiredis-client-0.16.0`. Are you using hiredis-client as well?

@andrcuns Any clue on how to debug this issue would be highly appreciated.

**Comment by rakibulislam** [comment]
> 

I can see the sidekiq-alive is starting successfully:
```
{"ts":"2023-09-04T14:02:25.641Z","pid":1,"tid":"3qzh","lvl":"INFO","msg":"Starting sidekiq-alive: {:hostname=>\"website-sidekiq-564f44bf99-k29cc\", :port=>7433, :ttl=>600, :queue=>\"sidekiq-alive-website-sidekiq-564f44bf99-k29cc\", :register_set=>\"sidekiq-alive-hostnames\", :liveness_key=>\"SIDEKIQ::LIVENESS_PROBE_TIMESTAMP::website-sidekiq-564f44bf99-k29cc\", :register_key=>\"SIDEKIQ_REGISTERED_INSTANCE::website-sidekiq-564f44bf99-k29cc\"}"}
169
{"ts":"2023-09-04T14:02:25.648Z","pid":1,"tid":"3qzh","lvl":"INFO","msg":"Successfully started sidekiq-alive, registered with key: SIDEKIQ_REGISTERED_INSTANCE::website-sidekiq-564f44bf99-k29cc on set sidekiq-alive-hostnames"}
```

I will try removing hiredis and deploy, and see if that fixes the issue. Will reply back once I test it out. Thanks so much for your help thius far @andrcuns :-)

**Comment by andrcuns** [comment]
If it starts successfully, the issue is then either with the job not being scheduled which is supposed to update alive key, the job being scheduled but being unsuccessful at updating the key. I believe it's one of these two.

The fact that it starts failing only after 10 minutes + the log shows that it creates liveness key successfully.

**Comment by rakibulislam** [comment]
@andrcuns So, I tried removing hiredis client but still the issue persists. I can see the sidekiq alive queue here: https://app.com.com/sidekiq/queues but it does not appear in the scheduled tab: https://app.com/sidekiq/scheduled that means the job is not being scheduled. Do you have any idea where I can look deep into in order to see what's the reason the job is not being scheduled? Thank you once again!

**Comment by andrcuns** [comment]
@rakibulislam Not really sure. I imagine you can try enabling `debug` logging for sidekiq and see if that reveals anything.

We could probably use some additional debug log messages in this repo as well, like actions being performed during startup and inside worker.

If it's not getting scheduled, I'm guessing it means either [this](https://github.com/arturictus/sidekiq_alive/blob/master/lib/sidekiq_alive.rb#L36) line did not schedule the initial job or it did but [this](https://github.com/arturictus/sidekiq_alive/blob/master/lib/sidekiq_alive/worker.rb#L16) action did not trigger the next job

**Comment by rakibulislam** [comment]
@andrcuns I was debugging this locally and found out that, [this line](https://github.com/arturictus/sidekiq_alive/blob/master/lib/sidekiq_alive.rb#L36) is executed. But, [this line](https://github.com/arturictus/sidekiq_alive/blob/master/lib/sidekiq_alive/worker.rb#L16) is not executed automatically. And, I think that's the main issue we need to figure out. But, when I run this command directly from the rails console: `SidekiqAlive::Worker.perform_async(hostname)` then, it runs but gets this error `ConnectionPool::TimeoutError`:
```
{"ts":"2023-09-13T04:33:33.682Z","pid":45635,"tid":"3f5v","lvl":"INFO","msg":"start","ctx":{"class":"SidekiqAlive::Worker","jid":"6e47fd68626fdb61325b0445"}}
SidekiqAlive::Worker.perform begins: 1
SidekiqAlive::Worker.perform begins: 2
****************************************************************************************************
store_alive_key!!! current_lifeness_key: SIDEKIQ::LIVENESS_PROBE_TIMESTAMP::HOSTNAME_NOT_SET
****************************************************************************************************
{"ts":"2023-09-13T04:33:34.791Z","pid":45635,"tid":"3f5v","lvl":"INFO","msg":"fail","ctx":{"class":"SidekiqAlive::Worker","jid":"6e47fd68626fdb61325b0445","elapsed":1.109}}
{"ts":"2023-09-13T04:33:34.792Z","pid":45635,"tid":"3f5v","lvl":"WARN","msg":"{\"context\":\"Job raised exception\",\"job\":{\"retry\":false,\"queue\":\"default\",\"args\":[\"HOSTNAME_NOT_SET\"],\"class\":\"SidekiqAlive::Worker\",\"jid\":\"6e47fd68626fdb61325b0445\",\"created_at\":1694579613.643147,\"enqueued_at\":1694579613.646084,\"error_message\":\"Waited 1 sec, 0/2 available\",\"error_class\":\"ConnectionPool::TimeoutError\",\"processor\":\"Ks-MacBook-Pro.local:45635\",\"failed_at\":1694579614.7854972},\"_config\":{\"labels\":\"#<Set: {}>\",\"require\":\".\",\"environment\":\"development\",\"concurrency\":5,\"timeout\":25,\"poll_interval_average\":null,\"average_scheduled_poll_interval\":5,\"on_complex_arguments\":\"raise\",\"error_handlers\":[\"#<Proc:0x0000000105925350 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq-7.1.2/lib/sidekiq/config.rb:37 (lambda)>\",\"#<Raven::SidekiqErrorHandler:0x0000000108595d18>\"],\"death_handlers\":[\"#<Proc:0x00000001096d8170 /Users/k.m.rakibulislam/Development/Yoshi/website/config/initializers/sidekiq.rb:9 (lambda)>\",\"#<Proc:0x000000010a6c92c8 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq-unique-jobs-8.0.3/lib/sidekiq_unique_jobs/server.rb:8 (lambda)>\"],\"lifecycle_events\":{\"startup\":[],\"quiet\":[\"#<Proc:0x000000010a34a070 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq_alive-2.3.0/lib/sidekiq_alive.rb:45>\"],\"shutdown\":[\"#<Proc:0x000000010a34a048 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq_alive-2.3.0/lib/sidekiq_alive.rb:50>\",\"#<Proc:0x00000001096d7b08 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq-unique-jobs-8.0.3/lib/sidekiq_unique_jobs/server.rb:21>\"],\"heartbeat\":[],\"beat\":[\"#<Proc:0x000000010546b328 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq-7.1.2/lib/sidekiq/metrics/tracking.rb:133>\"]},\"dead_max_jobs\":10000,\"dead_timeout_in_seconds\":15552000,\"reloader\":\"#<Sidekiq::Rails::Reloader @app=Website::Application>\",\"backtrace_cleaner\":\"#<Proc:0x000000010a5693d8 /Users/k.m.rakibulislam/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/sidekiq-7.1.2/lib/sidekiq/rails.rb:59 (lambda)>\",\"pidfile\":\"tmp/pids/sidekiq.pid\",\"beta\":{\"concurrency\":5},\"staging\":{\"concurrency\":5},\"production\":{\"concurrency\":20},\"queues\":[\"charging\",\"high_pri\",\"order_positioning\",\"delivery_positioning\",\"membership_renew\",\"messaging\",\"mailers\",\"default\",\"low_pri\",\"set_first_effective_eta\",\"elasticsearch\"],\"limits\":{\"delivery_positioning\":1},\"logfile\":\"log/sidekiq.log\",\"config_file\":\"./config/sidekiq.yml\",\"tag\":\"website\",\"identity\":\"Ks-MacBook-Pro.local:45635:cd297817bf8b\",\"cron_poll_interval\":30,\"fetch_class\":\"Sidekiq::LimitFetch\"}}"}
{"ts":"2023-09-13T04:33:34.793Z","pid":45635,"tid":"3f5v","lvl":"WARN","msg":"ConnectionPool::TimeoutError: Waited 1 sec, 0/2 available"}
```

Do you see why this might be happening? Thanks for your time and help!

**Comment by rakibulislam** [comment]
@andrcuns Looks like its enqueuing the `SidekiqAlive::Worker` job but it's not being processed so nothing is scheduled. But, if I run the same command: `SidekiqAlive::Worker.perform_async(hostname)` from the rails console, it does enqueue and then gets processed and hence it shows up in the scheduled tab in the sidekiq UI. Any thoughts on what could be happening here?

**Comment by Javierchik** [comment]
Hey Team!
`Can't find the alive key` message is also showing to me in an Azure web app (service) that runs Sidekiq 6.5.9
I was able to reproduce the exact same behavior, unhealthy after 10 minutes, in my local.
It looks to me that somehow Sidekiq does not get the reference to the new queue that is created after 10min (default is 10 min)

![Screenshot 2023-09-15 at 8 09 10 AM](https://github.com/arturictus/sidekiq_alive/assets/242332/6ef7a2ba-1f89-4843-b715-b8cf04b1359c)

Appreciate any help!

**Comment by rakibulislam** [comment]
Okay, I fixed this issue, finally! For me everything was working fine with sidekiq 6.5.6 but the sidekiq-alive stopped processing and scheduling jobs after we upgraded to sidekiq 7.1. Finally, the solution was to add the sidekiq-alive queue in the `sidekiq,yml` file and I defined the queue name this way: `<%= "sidekiq-alive-#{ENV["HOSTNAME"] || 'HOSTNAME_NOT_SET'}" %>`. After this, sidekiq-alive started working as expected with sidekiq 7.1!

**Comment by andrcuns** [comment]
That's not really a fix though, queue should be created automatically. In the latest version with sidekiq 7+ it uses a separate capsule as well.

It would be nice to understand why the queue was not there.

**Comment by rakibulislam** [comment]
@andrcuns I agree with you. It would be great to understand why it is not creating the sidekiq-alive queue automatically with sidekiq 7.1. But, the fix works for us for now.

**Comment by w32-blaster** [comment]
This really sound like a workaround to me, but @rakibulislam did a great work to figure this out. 
Wondering if a good approach and next step would be to add debug level logs to sidekiq-alive, so the actual root cause can be found or at least narrowed down.

**Comment by andrcuns** [comment]
Yeah, some additional debug level messages wouldn't hurt. I'll see if I can add a pr sometime this week

**Comment by arturictus** [comment]
Hi there,
Trying to figure out here what could be the cause.
1. queues not being added automatically:
In @Javierchik picture looks like the queues are set: https://github.com/arturictus/sidekiq_alive/issues/94#issuecomment-1721259607
But there are still doubts about that. 
__QUESTION:__ could please confirm if the queues are listed in web or in the instance?

2. Hostname not set:
In case you are running multiple instances they will have the same redis key which will cause the opposite effect, return alive with blocked or inactive instances. 
__This library is intended to run with the HOSTNAME present_

3. Worker not being scheduled:
Looks improbable since these logs show success and the queuing is done before success: https://github.com/arturictus/sidekiq_alive/issues/94#issuecomment-1705354958

4. [Error in job connection pool](https://github.com/arturictus/sidekiq_alive/issues/94#issuecomment-1716933190):
- [ ] We should retry the job in case this happens -> 🗒️ 

5. Suspicion :
Maybe sidekiq 7 loads/copies the queues in memory before we run `on(:startup)` effectively not running our queue but being visible on the web
Best solution I can think right now to that is the `sidekiq.yml` option that @rakibulislam proposed [here](https://github.com/arturictus/sidekiq_alive/issues/94#issuecomment-1722111354)

6. TODO:
- [ ] Review the capsule functionality just in case we missed something in the implementation.

__QUESTION:_ are you running redis in cluster? or single instance?, replicas?

**Comment by w32-blaster** [comment]
@arturictus we are running redis in Elasticache as a single instance. Not in cluster mode. There are no replicas currently

**Comment by breiko83** [comment]
Facing the same issue here. The 10 mins mark is actually based on the default TTL for Redis key in the config file:

```
  # ==> Time to live
  # Time for the key to be kept by Redis.
  # Here is where you can set de periodicity that the Sidekiq has to probe it is working
  # Time unit: seconds
  # default: 10 * 60 # 10 minutes
```

I reduced this to 60 and the worker goes unhealthy after 1 minute (Can't find the alive key)

**Comment by amkisko** [comment]
Confirming the issue, exactly after 10 minutes worker gets unhealthy status, so basically sidekiq_alive returns "not alive", after than sidekiq starts failing with `ConnectionPool::TimeoutError: Waited 1 sec, 0/2 available` which seems to be relevant to missing workers.

**Comment by jasondoc3** [comment]
I'm not sure if this will help anyone else, but I was also experiencing this problem with sidekiq `6.5.7`. After some deep debugging, I discovered that the issue for me was actually the [sidekiq-throttled](https://github.com/ixti/sidekiq-throttled) gem.

On the throttled [README](https://github.com/ixti/sidekiq-throttled#usage), it states:
> Load order can be an issue if you are using other Sidekiq plugins and/or middleware. To prevent any problems, add the .setup! call to the bottom of your init file.

Because `sidekiq_alive` uses an `on(:startup)` lifecycle event to modify the configuration, I modified my sidekiq initializer to setup the throttled gem in a separate `on(:startup)` lifecycle event to correct the load order:

```rb
# config/initializers/sidekiq.rb
require "sidekiq/throttled"
Sidekiq.configure_server do |config|
  config.on(:startup) { Sidekiq::Throttled.setup! }
end
```

**Comment by mleiseca** [comment]
@amkisko I was having the same issue as your comment [here](https://github.com/arturictus/sidekiq_alive/issues/94#issuecomment-1750010859)

1. SidekiqAlive.start executes and calls SidekiqAlive::Worker.perform_async
2. When SidekiqAlive::Worker.perform executes, it calls `store_alive_key`. This fails with
```
ConnectionPool::TimeoutError: Waited 1 sec, 0/2 available
```

Fix:

In `Sidekiq.configure_server` add a section that creates a capsule for SidekiqAlive
```
Sidekiq.configure_server do |config|

  config.capsule(SidekiqAlive::CAPSULE_NAME) do |cap|
    cap.concurrency = 3
    hostname = ENV["HOSTNAME"] || "HOSTNAME_NOT_SET"
    cap.queues = ["sidekiq-alive-#{hostname}"]
  end
end
```

I don't really know anything about Sidekiq internals, but it seems like it's creating a Redis connection pool that's too small - somehow there are only two connections, but the pool is exhausted when this method is called:

```
    def store_alive_key
      redis.set(current_lifeness_key, time: Time.now.to_i, ex: config.time_to_live.to_i)
    end
```

setting `cap.concurrency = 3` makes a slightly larger connection pool just for sidekiq alive

**Comment by stefanoc** [comment]
@jasondoc3 we were having the exact same problem with Sidekq 6.5.11 and luckily I found your comment. Thanks! :-)

**Comment by smoke** [comment]
I believe this is caused because of race condition when Sidekiq is starting up, it sets up `queues` in Redis
When I start more than 1 new Workers all at once - the `queues` is not having each and every new worker that started.

For example I started 3 workers all at once, and only 1 have managed to register its queue:
![image](https://github.com/user-attachments/assets/77618eb6-c6ec-4e18-be07-7e6694a5a308)


However if I start workers one by one, all is good. 
![image](https://github.com/user-attachments/assets/bd4a9f35-1361-49d9-92bd-a27fef66d72b)
